File: emarsys/src/main/java/com/emarsys/service/EmarsysMessagingServiceUtils.java
package com.emarsys.service;

import android.content.Context;

import com.emarsys.core.Callable;
import com.emarsys.core.di.DependencyContainer;
import com.emarsys.core.di.DependencyInjection;
import com.emarsys.core.util.Assert;
import com.emarsys.mobileengage.di.MobileEngageDependencyContainer;
import com.emarsys.mobileengage.service.MessagingServiceUtils;
import com.google.firebase.messaging.RemoteMessage;

public class EmarsysMessagingServiceUtils {

    public static boolean handleMessage(final Context context, final RemoteMessage remoteMessage) {
        Assert.notNull(context, "Context must not be null!");
        Assert.notNull(remoteMessage, "RemoteMessage must not be null!");

        final DependencyContainer container = DependencyInjection.<MobileEngageDependencyContainer>getContainer();

        return container.getRunnerProxy().logException(new Callable<Boolean>() {
            @Override
            public Boolean call() {
                return MessagingServiceUtils.handleMessage(
                        context,
                        remoteMessage,
                        container.getDeviceInfo());
            }
        });
    }

}

File: emarsys-firebase/src/main/java/com/emarsys/service/EmarsysMessagingServiceUtils.kt
package com.emarsys.service

import android.content.Context
import com.emarsys.core.handler.CoreSdkHandler
import com.emarsys.core.util.Assert
import com.emarsys.mobileengage.di.mobileEngage
import com.emarsys.mobileengage.service.MessagingServiceUtils
import com.emarsys.mobileengage.service.MessagingServiceUtils.handleMessage
import com.google.firebase.messaging.RemoteMessage

object EmarsysMessagingServiceUtils {
    var MESSAGE_FILTER = MessagingServiceUtils.MESSAGE_FILTER

    @JvmStatic
    fun handleMessage(context: Context, remoteMessage: RemoteMessage): Boolean {
        Assert.notNull(context, "Context must not be null!")
        Assert.notNull(remoteMessage, "RemoteMessage must not be null!")

        val handler: CoreSdkHandler = mobileEngage().coreSdkHandler

        handler.post {
            handleMessage(
                    context,
                    remoteMessage,
                    mobileEngage().deviceInfo,
                    mobileEngage().notificationCache,
                    mobileEngage().timestampProvider,
                    mobileEngage().fileDownloader,
                    mobileEngage().silentMessageActionCommandFactory,
                    mobileEngage().remoteMessageMapper)
        }

        return isMobileEngageMessage(remoteMessage.data)
    }

    fun isMobileEngageMessage(remoteMessageData: Map<String, String?>?): Boolean {
        return MessagingServiceUtils.isMobileEngageMessage(remoteMessageData)
    }
}